include/odp/api/spec/ipsec.h

```diff
@@ -1601,6 +1639,25 @@ int odp_ipsec_sa_mtu_update(odp_ipsec_sa_t sa, uint32_t mtu);
  */
 void *odp_ipsec_sa_context(odp_ipsec_sa_t sa);
 
+/**
+ * Get IPSEC stats for the IPSEC SA handle
+ *
+ * @param          sa       IPSEC SA handle
+ * @param[out]     stats    Stats output
+ *                          Stats returned will be 0 for invalid SA
+ */
+void odp_ipsec_stats(odp_ipsec_sa_t sa, odp_ipsec_stats_t *stats);
```

Comment:
Maybe both of these functions should return success status?

Reply:
Agreed. What would be better,
1. Return status (success/failure)
2. Return number of SAs processed (0 indicating failure)

Reply:
success/failure would seem more natural since the SAs do not form an ordered batch like packets do in other APIs. But it is still useful to be able to have the lookup fail for some SAs. So maybe it should work so that if return status is 0, then stats could be returned at least for some SAs (and others would show zeroes)?

Reply:
I've updated this in v3.

Reply:
I think return value of the stats_multi call should tell which statistics where successfully updated. Either all/nothing (0 vs -1), or N = number of stats successfully written (starting from stats[0]). Otherwise, application could not tell if stats of all zeros did not receive any packets, or if there was an error reading those stats.

Usually the call should always succeed (application gives valid SA handles). So, in practice the difference would be that retval N would report the first failing SA vs 0/-1 would just tell that some SA failed.

I'd vote retval N. It's consistent with other multi APIs (although here application would not retry a failing stats read) and has the extra ability to return (the first) failing SA.

Another option would be to add a success / failure flag into stats struct, which application could check if N != num. This would tell all failing SAs (normally N==num and no need to check the flag).

Reply:
IMHO, success/failure flag in the SA stats struct would be better than having the return value tell that the _first_ N SA stats were ok. The SAs are not ordered like packets are, so problematic SA *) in the beginning would prevent getting stats of the later SAs unless the app reorders the SA request list. Such reordering or removal of packets in packet APIs happens naturally but not, in my view, for SAs in this SA stats case.

*) Actually, if the only failure case would be invalid SA handle, then we could just tell that invalid SA handles are not ok or have the whole function fail without returning anything until the application gets its act together. But maybe there could be other, transient, issues in fetching the stats (e.g. when SAs are being rapidly created and destroyed) so the success/failure flag could be useful.

Reply:
Initial plan was to have all stats to say 0 for an invalid SA. But looking at the discussion, I think it's better to fail the operation in such cases. I can remove the line,
`
Stats returned will be 0 for invalid SA`

And state that -EINVAL would be returned in case any SA is invalid. @psavol do you agree to this approach?

Reply:
Agree that this stats_multi call could just either succeed (return 0), or fail (<0). Application should pass only valid SAs to the call, and thus a failure should be very rare result.

For failure, specify <0 instead of a specific value (usually we return -1, but spec says <0).

Reply:
@psavol Can you take a look at latest patch? I had made this change (returning 0 for success & <0 for failure). I'll submit a new version to state that application should pass only valid SAs to this API.

Reply:
I do not think there is need to state that an application should pass only valid SAs to this API as that can be expected anyway. But the point is that if the call (single or multi-SA variant) succeeds, then stats are returned for all SAs. The possibility to return stats only for some SAs is thus removed, so these sentences should be removed: "Stats returned will be 0 for any invalid SA".

Reply:
@JannePeltonen Agreed. I'll make this change in next version. Shall I add your "Reviewed-by" tag in next version?

Reply:
Yes, please add the tag.